File: app/src/main/java/com/holokenmod/Grid.java
package com.holokenmod;

import com.holokenmod.options.ApplicationPreferences;

import java.util.ArrayList;
import java.util.Collection;

public class Grid {
	private final ArrayList<GridCell> cells = new ArrayList<>();
	private final ArrayList<GridCage> cages = new ArrayList<>();
	private final int mGridSize;
	private GridCell mSelectedCell;
	
	public Grid(final int gridSize) {
		this.mGridSize = gridSize;
	}
	
	public ArrayList<GridCell> getCells() {
		return cells;
	}
	
	public int getGridSize() {
		return mGridSize;
	}
	
	public GridCell getSelectedCell() {
		return mSelectedCell;
	}
	
	public void setSelectedCell(final GridCell SelectedCell) {
		this.mSelectedCell = SelectedCell;
	}
	
	public GridCage getCage(final int row, final int column) {
        if (row < 0 || row >= mGridSize || column < 0 || column >= mGridSize) {
            return null;
        }
		return cells.get(column + row * mGridSize).getCage();
	}
	
	public ArrayList<GridCell> invalidsHighlighted() {
		final ArrayList<GridCell> invalids = new ArrayList<>();
		for (final GridCell cell : cells) {
			if (cell.isInvalidHighlight()) {
				invalids.add(cell);
			}
		}
		
		return invalids;
	}
	
	public ArrayList<GridCell> cheatedHighlighted() {
		final ArrayList<GridCell> cheats = new ArrayList<>();
        for (final GridCell cell : cells) {
            if (cell.isCheated()) {
                cheats.add(cell);
            }
        }
		
		return cheats;
	}
	
	public void markInvalidChoices() {
		for (final GridCell cell : cells) {
			if (cell.isUserValueSet() && cell.getUserValue() != cell.getValue()) {
				cell.setInvalidHighlight(true);
			}
		}
	}
	
	// Returns whether the puzzle is solved.
	public boolean isSolved() {
        for (final GridCell cell : cells) {
            if (!cell.isUserValueCorrect()) {
                return false;
            }
        }
		return true;
	}
	
	public int countCheated() {
		int counter = 0;
        for (final GridCell cell : cells) {
            if (cell.isCheated()) {
                counter++;
            }
        }
		return counter;
	}
	
	public int[] countMistakes() {
		final int[] counter = {0, 0};
		for (final GridCell cell : cells) {
			if (cell.isUserValueSet()) {
				counter[1]++;
                if (cell.getUserValue() != cell.getValue()) {
                    counter[0]++;
                }
			}
		}
		return counter;
	}
	
	/* Clear any cells containing the given number. */
	public void clearValue(final int value) {
        for (final GridCell cell : cells) {
            if (cell.getValue() == value) {
                cell.setValue(-1);
            }
        }
	}
	
	/* Determine if the given value is in the given column */
	public boolean valueInColumn(final int column, final int value) {
        for (int row = 0; row < mGridSize; row++) {
            if (cells.get(column + row * mGridSize).getValue() == value) {
                return true;
            }
        }
		return false;
	}
	
	// Return the number of times a given user value is in a row
	public int getNumValueInRow(final GridCell ocell) {
		int count = 0;
        for (final GridCell cell : cells) {
            if (cell.getRow() == ocell.getRow() &&
                    cell.getUserValue() == ocell.getUserValue()) {
                count++;
            }
        }
		return count;
	}
	
	// Return the number of times a given user value is in a column
	public int getNumValueInCol(final GridCell ocell) {
		int count = 0;
        
        for (final GridCell cell : cells) {
            if (cell.getColumn() == ocell.getColumn() &&
                    cell.getUserValue() == ocell.getUserValue()) {
                count++;
            }
        }
		return count;
	}
	
	// Return the cells with same possibles in row and column
	public ArrayList<GridCell> getPossiblesInRowCol(final GridCell ocell) {
		final ArrayList<GridCell> possiblesRowCol = new ArrayList<>();
		final int userValue = ocell.getUserValue();
        for (final GridCell cell : cells) {
            if (cell.isPossible(userValue)) {
                if (cell.getRow() == ocell.getRow() || cell.getColumn() == ocell.getColumn()) {
                    possiblesRowCol.add(cell);
                }
            }
        }
		return possiblesRowCol;
	}
	
	public ArrayList<GridCell> getSinglePossibles() {
		final ArrayList<GridCell> singlePossibles = new ArrayList<>();
        for (final GridCell cell : cells) {
            if (cell.getPossibles().size() == 1) {
                singlePossibles.add(cell);
            }
        }
		return singlePossibles;
	}
	
	public GridCell getCellAt(final int row, final int column) {
		if (!isValidCell(row, column)) {
			return null;
		}
		
		return cells.get(column + row * mGridSize);
	}
	
	public boolean isValidCell(final int row, final int column) {
		return row >= 0 && row < mGridSize
				&& column >= 0 && column < mGridSize;
	}
	
	public ArrayList<GridCage> getCages() {
		return cages;
	}
	
	public void ClearAllCages() {
		for (final GridCell cell : this.cells) {
			cell.setCage(null);
			cell.setCagetext("");
		}
		this.cages.clear();
	}
	
	public void setCageTexts() {
		for (final GridCage cage : cages) {
		    cage.updateCageText();
		}
	}
	
	public void addCell(final GridCell cell) {
		this.cells.add(cell);
	}
	
	public GridCell getCell(final int index) {
		return this.cells.get(index);
	}
	
	public void addCage(final GridCage cage) {
		this.cages.add(cage);
	}
	
	public void clearUserValues() {
		for (final GridCell cell : cells) {
			cell.clearUserValue();
			cell.setCheated(false);
		}
		
		if (mSelectedCell != null) {
			mSelectedCell.setSelected(false);
			mSelectedCell.getCage().setSelected(false);
		}
	}
	
	public void clearLastModified() {
		for (final GridCell cell : cells) {
			cell.setLastModified(false);
		}
	}
	
	public void solve(final boolean solveGrid) {
		if (mSelectedCell != null) {
			ArrayList<GridCell> solvecell = mSelectedCell.getCage().getCells();
			if (solveGrid) {
				solvecell = new ArrayList<>(cells);
			}
			
			for (final GridCell cell : solvecell) {
				if (!cell.isUserValueCorrect()) {
					cell.setUserValueIntern(cell.getValue());
					cell.setCheated(true);
				}
			}
			mSelectedCell.setSelected(false);
			mSelectedCell.getCage().setSelected(false);
		}
	}
	
	public Collection<Integer> getPossibleDigits() {
		return ApplicationPreferences
				.getInstance()
				.getDigitSetting()
				.getPossibleDigits(mGridSize);
	}
	
	public int getMaximumDigit() {
		return ApplicationPreferences
				.getInstance()
				.getDigitSetting()
				.getMaximumDigit(mGridSize);
	}
	
	public Collection<Integer> getPossibleNonZeroDigits() {
		return ApplicationPreferences
				.getInstance()
				.getDigitSetting()
				.getPossibleNonZeroDigits(mGridSize);
	}
}

File: gauguin-core/src/main/kotlin/org/piepmeyer/gauguin/grid/Grid.kt
package org.piepmeyer.gauguin.grid

import io.github.oshai.kotlinlogging.KotlinLogging
import org.piepmeyer.gauguin.options.GameOptionsVariant
import org.piepmeyer.gauguin.options.GameVariant
import org.piepmeyer.gauguin.undo.UndoStep
import kotlin.time.Duration

private val logger = KotlinLogging.logger {}

class Grid(
    val variant: GameVariant,
) {
    val cells = createCells()
    var cages: List<GridCage> = mutableListOf()
    var selectedCell: GridCell? = null

    var playTime: Duration = Duration.ZERO
    var solvedFirstTimeOfKind = false
    var solvedBestTimeOfKind = false
    var description: String? = null

    var startedToBePlayed = false
    var isActive = false
    var creationDate: Long = 0
        private set

    val undoSteps = mutableListOf<UndoStep>()

    constructor(variant: GameVariant, creationDate: Long) : this(variant) {
        this.creationDate = creationDate
    }

    private fun createCells(): List<GridCell> {
        val listOfCells = mutableListOf<GridCell>()

        var cellnum = 0
        for (row in 0 until variant.height) {
            for (column in 0 until variant.width) {
                listOfCells += GridCell(cellnum++, row, column)
            }
        }

        return listOfCells.toList()
    }

    val gridSize: GridSize
        get() = variant.gridSize

    fun getCage(
        row: Int,
        column: Int,
    ): GridCage? =
        if (!isValidCell(row, column)) {
            null
        } else {
            cells[column + row * variant.width].cage
        }

    fun invalidsHighlighted(): List<GridCell> =
        cells.filter {
            it.isInvalidHighlight
        }

    fun cheatedHighlighted(): List<GridCell> =
        cells.filter {
            it.isCheated
        }

    fun markInvalidChoices() {
        cells.forEach { it.isInvalidHighlight = it.shouldBeHighlightedInvalid() }
    }

    fun isSolved(): Boolean = !cells.any { !it.isUserValueCorrect }

    fun numberOfMistakes(): Int {
        logger.info { "Calculating number of mistakes of:" }
        logger.info { detailedToString() }

        val mistakes = cells.count { it.isUserValueSet && !it.isUserValueCorrect }

        logger.info { "Counted mistakes: $mistakes" }

        return mistakes
    }

    private fun getNumValueInRow(ocell: GridCell): Int =
        cells.count {
            it.row == ocell.row &&
                it.userValue == ocell.userValue
        }

    private fun getNumValueInCol(ocell: GridCell): Int =
        cells.count {
            it.column == ocell.column &&
                it.userValue == ocell.userValue
        }

    fun getPossiblesInRowCol(cell: GridCell): List<GridCell> =
        cells
            .filter { it.isPossible(cell.userValue) }
            .filter { it.row == cell.row || it.column == cell.column }

    fun setUserValueAndRemovePossibles(
        cell: GridCell,
        value: Int,
    ) {
        cell.setUserValueExtern(value)

        removePossiblesFromCellValue(cell)
    }

    private fun removePossiblesFromCellValue(selectedCell: GridCell) {
        getPossiblesInRowCol(selectedCell).forEach {
            it.removePossible(selectedCell.userValue)
        }
    }

    fun getCellAt(
        row: Int,
        column: Int,
    ): GridCell? {
        if (!isValidCell(row, column)) {
            return null
        }

        return getValidCellAt(row, column)
    }

    fun getValidCellAt(
        row: Int,
        column: Int,
    ): GridCell = cells[column + row * variant.width]

    private fun isValidCell(
        row: Int,
        column: Int,
    ): Boolean = row >= 0 && row < variant.height && column >= 0 && column < variant.width

    fun clearAllCages() {
        for (cell in cells) {
            cell.cage = null
        }
        cages = mutableListOf()
    }

    fun getCell(index: Int): GridCell = cells[index]

    fun addCage(cage: GridCage) {
        cages = cages + cage
    }

    fun clearUserValues() {
        for (cell in cells) {
            cell.clearUserValue()
            cell.isCheated = false
        }

        selectedCell?.let {
            it.isSelected = false
        }
    }

    fun clearLastModified() {
        for (cell in cells) {
            cell.isLastModified = false
        }
    }

    override fun toString(): String = GridToString(this).printGrid()

    fun detailedToString(): String = super.toString() + " - " + toString()

    fun isUserValueUsedInSameRow(
        cellIndex: Int,
        value: Int,
    ): Boolean {
        val startIndex = cellIndex - cellIndex % variant.width
        for (index in startIndex until startIndex + variant.width) {
            if (index != cellIndex && cells[index].userValue == value) {
                return true
            }
        }
        return false
    }

    fun isUserValueUsedInSameColumn(
        cellIndex: Int,
        value: Int,
    ): Boolean {
        var index = cellIndex % variant.width
        while (index < variant.surfaceArea) {
            if (index != cellIndex && cells[index].userValue == value) {
                return true
            }
            index += variant.width
        }
        return false
    }

    fun isValueUsedInSameRow(
        cellIndex: Int,
        value: Int,
    ): Boolean {
        val startIndex = cellIndex - cellIndex % variant.width
        for (index in startIndex until startIndex + variant.width) {
            if (index != cellIndex && cells[index].value == value) {
                return true
            }
        }
        return false
    }

    fun isValueUsedInSameColumn(
        cellIndex: Int,
        value: Int,
    ): Boolean {
        var index = cellIndex % variant.width
        while (index < variant.surfaceArea) {
            if (index != cellIndex && cells[index].value == value) {
                return true
            }
            index += variant.width
        }
        return false
    }

    fun addPossiblesAtNewGame() {
        cells.forEach { it.possibles = variant.possibleDigits }
    }

    fun userValueChanged() {
        updateDuplicatedNumbersInRowOrColumn()
    }

    fun updateDuplicatedNumbersInRowOrColumn() {
        cells.forEach {
            it.duplicatedInRowOrColumn = it.isUserValueSet && (getNumValueInCol(it) > 1 || getNumValueInRow(it) > 1)
        }
    }

    fun isCheated(): Boolean = cells.any { it.isCheated }

    fun hasCellsWithSinglePossibles(): Boolean = cells.any { it.possibles.size == 1 }

    fun getCellsAtSameRow(cell: GridCell): List<GridCell> {
        return cells.filter { it.row == cell.row && it != cell }
    }

    fun getCellsAtSameColumn(cell: GridCell): List<GridCell> {
        return cells.filter { it.column == cell.column && it != cell }
    }

    val options: GameOptionsVariant
        get() = variant.options

    fun belongsCellToTheEastOfFirstCellToSameCage(
        cage: GridCage,
        distance: Int,
    ): Boolean {
        val cellToTheEast =
            getCellAt(
                cage.cells.first().row,
                cage.cells.first().column + distance,
            )

        return cellToTheEast?.cage == cage
    }

    fun areAdjacent(
        firstCage: GridCage,
        secondCage: GridCage,
    ): Boolean =
        firstCage.cells.any { cell ->
            getCage(cell.row + 1, cell.column) == secondCage ||
                getCage(cell.row - 1, cell.column) == secondCage ||
                getCage(cell.row, cell.column + 1) == secondCage ||
                getCage(cell.row, cell.column - 1) == secondCage
        }
}
